### PR TITLE
Remove more unnecessary mutex (based on #441)

### PR DESCRIPTION
**Pull request by garbageslam**
Remove unnecessary mutex around peer_keepalive
    
    This follows up on the commit that removes the mutex around
    byzantine ledger. The peer_keepalive mutex is only there to allow
    stopping the service. We change this from `Arc<Mutex<PeerKeepalive>`
    to `Option<Arc<PeerKeepalive>>`, and replace to call to `lock().stop()`
    in the stop function with `= None` which simply drops the value.
    
    We also make the other Arc's that are holding this, in lambda functions,
    into weak references. This prevents strong reference cycles which would
    prevent the drop from happening and cause a resource leak.
    
    We also put a code comment about why not to remove the mutex around
    threaded broadcaster.

### CODE REVIEW

**Comment by garbageslam** [comment]
rebased on master (after 441 merge)

File: consensus/service/src/consensus_service.rs
Comment (mfaulk) [inline consensus/service/src/consensus_service.rs:485]: Is there a situation where consensus_service would be running after the peer_keepalive has been dropped? I'm wondering if `upgrade().expect()` would be better than `if let Some`.

**Review by mfaulk — APPROVED** [review]

File: consensus/service/src/consensus_service.rs
Comment (garbageslam) [reply consensus/service/src/consensus_service.rs:486]: yeah good call, that seems reasonable

**Review by eranrund — APPROVED** [review]
LGTM once k8 passes.

File: consensus/service/src/consensus_service.rs
Comment (garbageslam) [reply consensus/service/src/consensus_service.rs:486]: @mfaulk i guess the reason is, if someone calls `consensus_service.stop()` then this should get dropped, and that will cause the weak reference to return none. so if we panic when it becomes `none`, then the server will crash when we try to stop it

one way to deal with this might be, to try to change which of these objects are actually owning eachother. i think the only party that talks to the peer_keepalive variable may be, the `consensus_msgs_from_network` thread. so we could try to make that the exclusive owner, and just have it be an `Box` that gets moved into that thread

it seems to me that the main reason we have all these member variables of `consensus_service` is to support the `pub fn stop(&mut self)` thing, which stops all the threads without tearing down any of the objects. i'm not sure the `stop` thing is actually useful, it might be simpler to just stop the threads in the `drop` calls. That might reduce the amount of code overall, and in case like the peer_keepalive, it seems that we end up with mutexes only because we need exclusive access to stop the object, which doesn't seem like a good trade. `drop` has signature `drop(&mut self)` -- the language is ensuring that the call to the destructor is exclusive, so that's a good time to join things, IMO. that's just one opinion though, maybe there's a good use case for stopping without tearing down the objects.

**Comment by garbageslam** [comment]
this is deployed as diogenes/97: https://deploybot.canadacentral.mobilecoin.com/job/mobilecoin-internal/job/deploy%2Fdiogenes/97/console

**Comment by garbageslam** [comment]
very little change in slam perf:
pr443

**Pull request merged**
